### ISSUE TITLE
usbipd: error: WSL kernel is not USBIP capable

### ISSUE THREAD

**Issue by leochien1110** [dorssel/usbipd-win#580]
Hi,

I am trying to mount my webcam into wsl but having this error message:
`usbipd: error: WSL kernel is not USBIP capable; update with 'wsl --update'.`

Here are some info of my computer:
```
ver > Microsoft Windows [Version 10.0.22624.1546]

wsl --version > 
WSL version: 1.2.2.0
Kernel version: 5.15.90.1
WSLg version: 1.0.51
MSRDC version: 1.2.3770
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22624.1546

wsl --status >
Default Distribution: Ubuntu-20.04
Default Version: 2

```
I did follow the hint and run `wsl --update` and even shutdown the wsl, but the error still exists.

Does anyone encounter this issue and have any idea to fix it?

Thank you guys in advanced

**Comment by danielk-98** [comment]
Did you install USBIP support in Ubuntu? The guide is here in case you didn't see it: https://github.com/dorssel/usbipd-win/wiki/WSL-support

**Comment by leochien1110** [comment]
> Did you install USBIP support in Ubuntu? The guide is here in case you didn't see it: https://github.com/dorssel/usbipd-win/wiki/WSL-support

Yes, I ran these two lines in Ubuntu:
```
sudo apt install linux-tools-virtual hwdata
sudo update-alternatives --install /usr/local/bin/usbip usbip `ls /usr/lib/linux-tools/*/usbip | tail -n1` 20
```

**Comment by dorssel** [comment]
From within Windows, what is the output of `wsl --list --verbose`?
In your distro (Linux side), what is the output of `cat /sys/devices/platform/vhci_hcd.0/status`?

**Comment by fyoory** [comment]
This works for me catting that sysfs  status file but does not work to map USB from windows to WSL2.

`$ uname -a
Linux fury-nitro5 5.15.90.1-microsoft-standard-WSL2+ #2 SMP Fri Apr 14 12:02:50 CDT 2023 x86_64 x86_64 x86_64 GNU/Linux
$ usbipd -h
WARNING: usbipd not found for kernel 5.15.90.1-microsoft

  You may need to install the following packages for this specific kernel:
    linux-tools-5.15.90.1-microsoft-standard-WSL2+
    linux-cloud-tools-5.15.90.1-microsoft-standard-WSL2+

  You may also want to install one of the following packages to keep up to date:
    linux-tools-standard-WSL2+
    linux-cloud-tools-standard-WSL2+

$ zgrep ^CONFIG_USB= /proc/config.gz
CONFIG_USB=y`

This kernel was rebuilt following exactly using https://github.com/dorssel/usbipd-win/wiki/WSL-support#building-your-own-usbip-enabled-wsl-2-kernel.

Throwing spaghetti at the wall here but may be that the docs mention 5.10 but the latest on ubuntu 22.04 in WSL2 as of today uses the above 5.15.90.1 kernel.

**Comment by leochien1110** [comment]
> From within Windows, what is the output of `wsl --list --verbose`? In your distro (Linux side), what is the output of `cat /sys/devices/platform/vhci_hcd.0/status`?

In Windows:
```
>wsl --list --verbose
  NAME            STATE           VERSION
* Ubuntu-20.04    Running         2
```
In my Ubuntu, there's no such file:
```
~$ cat /sys/devices/platform/vhci_hcd.0/status
cat: /sys/devices/platform/vhci_hcd.0/status: No such file or directory
```
This might be the reason.

Here's extra info on my WSL:
```
~$ uname -r
5.10.102.1-microsoft-standard-WSL2+
```

**Comment by leochien1110** [comment]
> This works for me catting that sysfs status file but does not work to map USB from windows to WSL2.
> 
> ```
> $ uname -a
> Linux fury-nitro5 5.15.90.1-microsoft-standard-WSL2+ #2 SMP Fri Apr 14 12:02:50 CDT 2023 x86_64 x86_64 x86_64 GNU/Linux
> $ usbipd -h
> WARNING: usbipd not found for kernel 5.15.90.1-microsoft
> 
>   You may need to install the following packages for this specific kernel:
>     linux-tools-5.15.90.1-microsoft-standard-WSL2+
>     linux-cloud-tools-5.15.90.1-microsoft-standard-WSL2+
> 
>   You may also want to install one of the following packages to keep up to date:
>     linux-tools-standard-WSL2+
>     linux-cloud-tools-standard-WSL2+
> 
> $ zgrep ^CONFIG_USB= /proc/config.gz
> CONFIG_USB=y
> ```
> 
> This kernel was rebuilt following exactly using https://github.com/dorssel/usbipd-win/wiki/WSL-support#building-your-own-usbip-enabled-wsl-2-kernel.
> 
> Throwing spaghetti at the wall here but may be that the docs mention 5.10 but the latest on ubuntu 22.04 in WSL2 as of today uses the above 5.15.90.1 kernel.

This is my result in the Ubuntu:
```
:~$ usbipd -h
WARNING: usbipd not found for kernel 5.10.102.1-microsoft

  You may need to install the following packages for this specific kernel:
    linux-tools-5.10.102.1-microsoft-standard-WSL2+
    linux-cloud-tools-5.10.102.1-microsoft-standard-WSL2+

  You may also want to install one of the following packages to keep up to date:
    linux-tools-standard-WSL2+
    linux-cloud-tools-standard-WSL2+
```
but I couldn't apt install the packages it mentioned. 
BTW, I clean-installed the wsl again yesterday but got the same error. Just wondering if someone is going to run Ubuntu20.04 might face the same issue.

**Comment by dorssel** [comment]
So you are using a self-built kernel? Looks like you did not enable USBIP. Does your /proc/config.gz have:
```
CONFIG_USBIP_CORE=y
CONFIG_USBIP_VHCI_HCD=y
CONFIG_USBIP_VHCI_HC_PORTS=8
CONFIG_USBIP_VHCI_NR_HCS=1
```

And this:
```
:~$ usbipd -h
WARNING: usbipd not found for kernel 5.10.102.1-microsoft
```
is totally normal. You are trying to run the *server* side (note de 'd' at the end of `usbipd`) within WSL. That's not supposed to work. You need the *client* part installed on WSL,; that's called `usbip` (no 'd' at the end).

**Comment by fyoory** [comment]
Yea those are fine,  Ok I will dig around more on my end as those are set too:

```
$ zgrep ^CONFIG_USBIP_ /proc/config.gz
CONFIG_USBIP_CORE=y
CONFIG_USBIP_VHCI_HCD=y
CONFIG_USBIP_VHCI_HC_PORTS=8
CONFIG_USBIP_VHCI_NR_HCS=1
```

**Comment by leochien1110** [comment]
What about `cat /sys/devices/platform/vhci_hcd.0/status` command having no such file? I don't think I miss any step from the instructions but `vhci_hcd` is still unavailable.

**Comment by dorssel** [comment]
> CONFIG_USBIP_CORE=m

You configured it as a module, instead of permanently enabled:

> CONFIG_USBIP_CORE=y

You could try `insmod`, but for WSL it is better to not modularize drivers.

**Comment by Xenfo** [comment]
I'm having the same issue, is there an issue with the prebuilt kernel on Ubuntu? Is the only way around this building my own?

**Comment by leochien1110** [comment]
> > CONFIG_USBIP_CORE=m
> 
> You configured it as a module, instead of permanently enabled:
> 
> > CONFIG_USBIP_CORE=y
> 
> You could try `insmod`, but for WSL it is better to not modularize drivers.

It is a new-installed wsl with nothing changed yet but just commands for usbip. Do you mean something like`insmod /proc/config.gz` to insert the module? Or is there any way to modify them from `m` to `y`?

**Comment by n9seo** [comment]
Hi,

I managed to get mine working fine on the kernel I mentioned and managed to attach a device to Ubuntu on 22.04 just fine.  Error was on my part.  So disregard my earlier reports.

**Comment by leochien1110** [comment]
So I tried the instruction [here](https://github.com/dorssel/usbipd-win/wiki/WSL-support#building-your-own-usbip-enabled-wsl-2-kernel) and checkout to linux-msft-wsl-5.15.y. I also follow the `menuconfig` to set the `CONFIG_USBIP_CORE` and `CONFIG_USBIP_VHCI_HCD` to `*`.
However, after rebooting, the result of `zgrep ^CONFIG_USBIP_ /proc/config.gz` is still showing:
```bash
CONFIG_USBIP_CORE=m
CONFIG_USBIP_VHCI_HCD=m
CONFIG_USBIP_VHCI_HC_PORTS=8
CONFIG_USBIP_VHCI_NR_HCS=1
```

**Comment by leochien1110** [comment]
> So I tried the instruction [here](https://github.com/dorssel/usbipd-win/wiki/WSL-support#building-your-own-usbip-enabled-wsl-2-kernel) and checkout to **linux-msft-wsl-5.15.y**. I also followed the `menuconfig` to set the `CONFIG_USBIP_CORE` and `CONFIG_USBIP_VHCI_HCD` to `*`. However, after rebooting, the result of `zgrep ^CONFIG_USBIP_ /proc/config.gz` is still showing:
> 
> ```shell
> CONFIG_USBIP_CORE=m
> CONFIG_USBIP_VHCI_HCD=m
> CONFIG_USBIP_VHCI_HC_PORTS=8
> CONFIG_USBIP_VHCI_NR_HCS=1
> ```
I FINALLY SOLVE IT!!!!

It looks like I missed some steps from the link I posted. After rerunning the whole instruction and building the kernel with `linux-msft-wsl-5.15.y` branch, the usbipd is now working on the Windows side and can be seen on the Ubuntu. I think that Microsoft's Ubuntu 20.04 is still running with kernel **5.10.102.1**(`uname -r`) which does not match the kernel version of WSL.
```bash
>wsl -v
WSL version: 1.2.4.0
Kernel version: 5.15.90.1
WSLg version: 1.0.51
MSRDC version: 1.2.3770
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22624.1610
```
In conclusion, updating the Ubuntu kernel version(now is **5.15.90.1-microsoft-standard-WSL2+**) to match the wsl kernel version is the solution. Microsoft always rolls out the new kernel several months later from the Microsoft Store.

Thanks to everyone who tried to help me get through this issue.

**Comment by DeeptangshuSaha** [comment]
I have two Distros one kali-linux and one Debian in the same machine.
But this problem only shows up in the Debian one. I tried all of the procedures shown here but none worked. I don't even know how to go about it anymore any ideas??

**Comment by dorssel** [comment]
What problem shows up in the Debian one? For USBIP, it really doesn't matter which distro is used for the attach, the device will be available in all distros, as there is only one WSL kernel. Are you using version 4.0.0?

**Comment by DeeptangshuSaha** [comment]
> What problem shows up in the Debian one? For USBIP, it really doesn't matter which distro is used for the attach, the device will be available in all distros, as there is only one WSL kernel. Are you using version 4.0.0?

Alright, So this is the output from my `wsl -v` command
```
WSL version: 2.0.14.0
Kernel version: 5.15.133.1-1
WSLg version: 1.0.59
MSRDC version: 1.2.4677
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22631.2861
```
and for `wsl --list --verbose`

```
  NAME          STATE           VERSION
* kali-linux    Running         2
  Debian        Running         2
```
Now When I try to attach a device this is what happens in `kali-linux`

```
➜ usbipd wsl attach --busid 1-2 --distribution kali-linux
```
and it connects the device to the kali but when I do this with Debian this is the output ➜

```
➜ usbipd wsl attach --busid 1-2 --distribution Debian
usbipd: error: WSL kernel is not USBIP capable; update with 'wsl --update'.
```
This is really confusing as to why this is happening. Please help!!

**Comment by dorssel** [comment]
Update to 4.0.0

**Comment by terlar** [comment]
I just ran into this issue. Just upgraded to:
```
PS > wsl --version
WSL version: 2.3.13.0
Kernel version: 6.6.36.3-1
WSLg version: 1.0.63
MSRDC version: 1.2.5326
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26100.1-240331-1435.ge-release
Windows version: 10.0.22631.3880
```

and 

```
 PS > usbipd --version
4.2.0+54.Branch.master.Sha.5b36605535dd52bc3d8a8e4bbaee7f4d480c86ad
```

I get the same issues and I see:
```
$ zgrep ^CONFIG_USBIP_ /proc/config.gz
CONFIG_USBIP_CORE=m
CONFIG_USBIP_VHCI_HCD=m
CONFIG_USBIP_VHCI_HC_PORTS=8
CONFIG_USBIP_VHCI_NR_HCS=1
CONFIG_USBIP_HOST=m
```

I wonder if something changed with the latest Kernel making it a module instead. I see this in the news-entry:
> Update Kernel to 6.6.36.3
> Hundreds of new kernel modules added.

**Comment by terlar** [comment]
I can confirm that downgrading to `2.2.4.0` solved the issue:
```
PS > wsl --version
WSL version: 2.2.4.0
Kernel version: 5.15.153.1-2
WSLg version: 1.0.61
MSRDC version: 1.2.5326
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26091.1-240325-1447.ge-release
Windows version: 10.0.22631.3880
```

```
$ zgrep ^CONFIG_USBIP_ /proc/config.gz
CONFIG_USBIP_CORE=y
CONFIG_USBIP_VHCI_HCD=y
CONFIG_USBIP_VHCI_HC_PORTS=8
CONFIG_USBIP_VHCI_NR_HCS=1
```

I'm not sure if it is the module part or something else with the new kernel.

**Comment by terlar** [comment]
The default config was changed here:
https://github.com/microsoft/WSL2-Linux-Kernel/commit/4d72cb5bb5d14e520500e84316a5b7e0f9e572cf#diff-14741777d782cbccd09de1b054da61cab2720f19f2386a61fad73534d765ba97L2742-R5663

**Comment by dorssel** [comment]
@terlar 
This has all been solved: https://github.com/dorssel/usbipd-win/issues/995
